stranding/stranding.py

```diff
@@ -0,0 +1,190 @@
+import logging
+import warnings
+
+from Bio.pairwise2 import align, format_alignment
+from Bio.Seq import Seq
+
+from seqseek import Chromosome
+
+from .exceptions import (MissingReferenceFlank,
+                        InconsistentAlignment,
+                        Unstrandable,
+                        FlanksTooShort)
+
+
+LOGGER = logging.getLogger(__name__)
+DEFAULT_MIN_FLANK_LENGTH = 15
+DEFAULT_WINDOW_EXTENSION = 0
+
+
+FORWARD_STRAND = 1
+REVERSE_STRAND = -1
+
+
+# empirically derived default values from stranding hundreds of thousands of flanks
+# from an Illumina beadchip. Two points are awarded for each matching base and one
+# point is subtracted for each mismatch. Gaps are strongly discouraged with a 5 point
+# penalty.
+DEFAULT_MATCH_SCORE = 2
+DEFAULT_MISMATCH_PENALTY = -1
+DEFAULT_GAP_OPEN_PENALTY = -5
+DEFAULT_TOLERANCE = 0.77
+
+
+class GenomeStranding(object):
+
+    def __init__(self,
+                 min_flank_length=DEFAULT_MIN_FLANK_LENGTH,
+                 tolerance=DEFAULT_TOLERANCE,
+                 match_score=DEFAULT_MATCH_SCORE,
+                 mismatch_penalty=DEFAULT_MISMATCH_PENALTY,
+                 gap_open_penalty=DEFAULT_GAP_OPEN_PENALTY):
+
+            self.min_flank_length = min_flank_length
+            self.tolerance = tolerance
+            self.match_score = match_score
+            self.mismatch_penalty = mismatch_penalty
+            self.gap_open_penalty = gap_open_penalty
+            if self.min_flank_length < DEFAULT_MIN_FLANK_LENGTH:
+                warnings.warn('Short flank lengths may lead to inaccurate alignments')
+
+    def is_high_scoring(self, score, query):
+        if len(query) < self.min_flank_length:
+            return False
+        return score > len(query) * self.match_score * self.tolerance
+
+    def is_perfect_score(self, score, query):
+        if len(query) < self.min_flank_length:
+            return False
+        return score == len(query) * self.match_score
+
+    def align(self, ref, query, score_only=True):
+        alignment = align.localms(ref, query, self.match_score, self.mismatch_penalty,
+                                  self.gap_open_penalty, self.mismatch_penalty,
+                                  score_only=score_only)
+        if score_only and not alignment:
+            return 0
+        return alignment
+
+    def align_and_log(self, ref, query):
+        alignments = self.align(ref, query, False)
+        for a in alignments:
+            if self.is_high_scoring(a, query):
```

Comment:
this is a bug. the consequence is that too many alignments may be logged because the tuple returned from biopython is compared to the integer score cutoff. fix forthcoming.